Task 3

azatdavliatshin opened the pull request:
**What was done?** 

- product-service is done: getProductsList  and getProductsById functions are implemented;

**Additional scope**

- [x] Async/await is used in lambda functions

- [x] ES6 modules are used for product-service implementation

- [x] Webpack is configured for product-service

- [x] SWAGGER documentation is created for product-service

- [x] Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) (You may use JEST)

- [x]  Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase.

- [x] Main error scenarios are handled by API ("Product not found" error, try catch blocks are used in lambda handlers).

**Link to product-service API:**
https://340ixzs2ig.execute-api.eu-west-1.amazonaws.com/dev/products
https://340ixzs2ig.execute-api.eu-west-1.amazonaws.com/dev/products/{productId}

**LInk to FE MR (YOUR OWN REPOSITORY)** - https://github.com/azatdavliatshin/the-art-vault-fe/pull/3

**swagger** - product-service/swagger.json 

Link to CloudFront: https://d2xkllu91rjhzi.cloudfront.net/

azatdavliatshin committed:
Task 3 Product API:
- add product service using serverless;
- add handler getProductList as ES module;
- add api ES module to emulate asynchronous actions;

azatdavliatshin committed:
Task 3 Product API:
- add model for product;

azatdavliatshin committed:
Task 3 Product API:
- resolve isse with CORS;

azatdavliatshin committed:
Task 3 Product API:
- add handler for get porduct by id API;

azatdavliatshin committed:
Task 3 Product API:
- add art data;

azatdavliatshin committed:
Task 3 Product API:
- add new property ro Product model;
- add links to images;

azatdavliatshin commented:
I hope I will handle SWAGGER and unit tests by the end of cross-check

azatdavliatshin committed:
Task 3 Product API:
- add serverless-aws-documentation plugin;
- add documentation properties to serverless.ts;
- swagger.json is generated;

generalov commented on product-service/api/index.ts:
```diff
@@ -0,0 +1,8 @@
+import products from './productList.json';
+
+export const getProductList = async (): Promise<Array<Product>> => Promise.resolve(products);
+export const getProductById = async (queryId: string): Promise<Product> => new Promise((resolve, reject) => {
```
What's the purpose of wrapping the body of the async function into a promise?

azatdavliatshin replied:
Just an imitation of asynchronous code, e.g. DB queries.

AmsterGet replied:
Why not just return the value from the `async` function? It returns the promise anyway.

azatdavliatshin replied:
Yep, you are absolutely right, I missed this point. 
But for me it's easier to remove async annotation

azatdavliatshin committed:
Task 3 Product API:
- add swagger documentation for 404 case of getProductById;
- update swagger.json

generalov commented on product-service/api/index.ts:
```diff
@@ -0,0 +1,8 @@
+export const getProductList = async (): Promise<Array<Product>> => Promise.resolve(products);
+export const getProductById = async (queryId: string): Promise<Product> => new Promise((resolve, reject) => {
+    const product = products.find(({id}) => id === queryId);
+    product ? resolve(product) : reject({message: 'Product not found!', error: new Error('Product not found!')});
```
For interoperability with other exceptions it is better to reject with an instanceof Error instead of plain object (`reject(new Error(...))`)

azatdavliatshin replied:
I've checked it and due to some reason error object is empty on AWS. Still don't know why

azatdavliatshin replied:
i.e. no trace, no message

azatdavliatshin commented:
I have resolved issue with swagger, product-service/swagger.json

azatdavliatshin commented:
Unit tests are left, I think, I will handle it too

generalov reviewed (approved):
LGTM

azatdavliatshin committed:
Task 3 Product API:
- add unit tests for getProductsList

azatdavliatshin commented:
I added tests for one handler, hopefully will be able to add to another one.

azatdavliatshin committed:
Task Product API:
- remove redundant async annotation

azatdavliatshin committed:
Task 3 Product API:
- ad tests for getProductById handler;

azatdavliatshin commented:
Last unit test is added
